### PR TITLE
Creating high level utility to sync layout data with cache policies (LayoutSyncManager)

### PR DESCRIPTION
**Pull request by bhariharan**
- Introduces a data model encapsulation for layout data.
- Introduces a new manager class that makes it easy to sync layout data and returns a strongly-typed object for ease of use. This manager takes care of indexing the data, storing and retrieving from `SmartStore`.
- Introduces 3 simple cache policies that can be specified while retrieving data.
- Tests for the new manager class.

### CODE REVIEW

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    /**
+     * Fetches layout data for the specified object type and layout type using the specified
+     * mode and triggers the supplied callback once complete.
+     *
+     * @param objectType Object type.
+     * @param layoutType Layout type. Defaults to "Full" if null is passed in.
+     * @param mode Fetch mode. See {@link Mode} for available modes.
+     * @param syncCallback Layout sync callback.
+     */
+    public void fetchLayout(String objectType, String layoutType, Mode mode,
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Public API that allows the caller to specify where to fetch data from based on cache policy passed in.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                }
+            }
+        } catch (Exception e) {
+            SmartSyncLogger.e(TAG, "Exception occurred while reading layout data from the cache", e);
+        }
+    }
+
+    private void onSyncComplete(String objectType, LayoutSyncCallback syncCallback, Layout layout) {
+        if (syncCallback != null) {
+            syncCallback.onSyncComplete(objectType, layout);
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Triggers callback once the operation is complete.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            smartStore.registerSoup(SOUP_NAME, INDEX_SPECS);
+        }
+    }
+
+    /**
+     * Callback interface for layout sync.
+     *
+     * @author bhariharan
+     */
+    public interface LayoutSyncCallback {
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Callback interface for completion of operation.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+    /**
+     * Enum for available data fetch modes.
+     *
+     * CACHE_ONLY - Fetches data from the cache and returns null if no data is available.
+     * CACHE_FIRST - Fetches data from the cache and falls back on the server if no data is available.
+     * SERVER_FIRST - Fetches data from the server and falls back on the cache if the server doesn't
+     * return data. The data fetched from the server is automatically cached.
+     */
+    public enum Mode {
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Cache modes.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/SyncManager.java [hunk 5]
```diff
@@ -137,7 +136,7 @@ public static synchronized SyncManager getInstance(UserAccount account, String c
      */
     public static synchronized SyncManager getInstance(UserAccount account, String communityId, SmartStore smartStore) {
         if (account == null) {
-            account = SmartStoreSDKManager.getInstance().getUserAccountManager().getCurrentUser();
+            account = SmartSyncSDKManager.getInstance().getUserAccountManager().getCurrentUser();
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/SyncManager.java]: Unrelated fix.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Represents the layout of a Salesforce object.
+ *
+ * @author bhariharan
+ * @see <a href="https://developer.salesforce.com/docs/atlas.en-us.uiapi.meta/uiapi/ui_api_responses_record_layout.htm">https://developer.salesforce.com/docs/atlas.en-us.uiapi.meta/uiapi/ui_api_responses_record_layout.htm</a>
+ */
+public class Layout {
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Data model of a layout object. Pretty much from the UI SDK public doc. I have added links here to that public doc. It's pretty well-defined.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+    private List<LayoutSection> sections;
+    private JSONObject rawData;
+
+    /**
+     * Creates an instance of this class from its JSON representation.
+     *
+     * @param object JSON object.
+     * @return Instance of this class.
+     */
+    public static Layout fromJSON(JSONObject object) {
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Just 1 public API to create an object from JSON.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+        }
+        return layout;
+    }
+
+    /**
+     * Returns the ID of this layout.
+     *
+     * @return ID of this layout.
+     */
+    public String getId() {
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Getters for the fields available.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/SalesforceObject.java [hunk 9]
```diff
@@ -36,9 +36,7 @@
  * This class represents a typical Salesforce object.
  *
  * @author bhariharan
- * @deprecated Will be removed in Mobile SDK 7.0.
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/SalesforceObject.java]: I had accidentally deprecated this class. This is meant to stick around. Only `SalesforceObjectType` goes away.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/target/LayoutSyncDownTarget.java [hunk 10]
```diff
@@ -93,6 +94,10 @@ public JSONArray startFetch(SyncManager syncManager, long maxTimeStamp) throws I
         final RestRequest request = RestRequest.getRequestForObjectLayout(syncManager.apiVersion,
                 objectType, layoutType);
         final RestResponse response = syncManager.sendSyncWithSmartSyncUserAgent(request);
+        final JSONObject responseJSON = response.asJSONObject();
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/target/LayoutSyncDownTarget.java]: Adding `objectType` while storing in `SmartStore`.

File: libs/test/SmartSyncTest/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManagerTest.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Tests for {@link LayoutSyncManager}.
+ *
+ * @author bhariharan
+ */
+@RunWith(AndroidJUnit4.class)
+@MediumTest
+public class LayoutSyncManagerTest extends ManagerTestCase {
```
Comment (bhariharan) [inline libs/test/SmartSyncTest/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManagerTest.java]: Test class for the new manager. Tests the different cache modes.

File: libs/test/SmartSyncTest/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+    private static final String COMPACT = "Compact";
+    private static final String ACCOUNT = "Account";
+
+    private LayoutSyncManager layoutSyncManager;
+    private LayoutSyncCallbackQueue layoutSyncCallbackQueue;
+
+    private static class LayoutSyncCallbackQueue implements LayoutSyncManager.LayoutSyncCallback {
```
Comment (bhariharan) [inline libs/test/SmartSyncTest/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManagerTest.java]: Blocking queue implementation to wait for callback.

File: libs/test/SmartSyncTest/src/com/salesforce/androidsdk/smartsync/manager/SyncManagerTest.java [hunk 13]
```diff
@@ -74,7 +74,7 @@
 import static java.util.Collections.singletonList;
 
 /**
- * Test class for SyncState.
+ * Test class for SyncManager.
```
Comment (bhariharan) [inline libs/test/SmartSyncTest/src/com/salesforce/androidsdk/smartsync/manager/SyncManagerTest.java]: Unrelated typo fix.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+ */
+public class LayoutSyncManager {
+
+    private static final String SOUP_NAME = "layouts";
+    private static final String FEATURE_LAYOUT_SYNC = "LY";
+    private static final String QUERY = "SELECT {" + SOUP_NAME + ":_soup} from {" + SOUP_NAME +
+            "} WHERE {" + SOUP_NAME + ":objectType} = '%s' and {" + SOUP_NAME + ":layoutType} = '%s'";
+    private static final String TAG = "LayoutSyncManager";
+    private static final IndexSpec[] INDEX_SPECS = new IndexSpec[] {
+        new IndexSpec("objectType", SmartStore.Type.string),
```
Comment (wmathurin) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Should we use type json1? Maybe at some point in the future we will deprecate other index types - or maybe we will need to add indexes to that table, when using json1, index changes can be done very quickly (because it's pure sql).

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+            final JSONArray results = smartStore.query(querySpec, 0);
+            if (results == null || results.length() == 0) {
+                if (fallbackOnServer) {
+                    fetchFromServer(objectType, layoutType, syncCallback);
+                } else {
+                    onSyncComplete(objectType, syncCallback, null);
+                }
+            } else {
+                final JSONArray resultArr = results.optJSONArray(0);
+                if (resultArr == null || resultArr.length() == 0) {
```
Comment (wmathurin) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: When would that happen? And shouldn't we then check fallbackOnServer.
In other words should those conditions move to line 228?

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+    public String getLayoutType() {
+        return layoutType;
+    }
+
+    /**
+     * Returns the mode of this layout.
+     *
+     * @return Mode of this layout.
+     */
+    public String getMode() {
```
Comment (wmathurin) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Should that return an enum type?

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+    public String getId() {
+        return id;
+    }
+
+    /**
+     * Returns the layout type of this layout.
+     *
+     * @return Layout type of this layout.
+     */
+    public String getLayoutType() {
```
Comment (wmathurin) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Should that return an enum type?

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/target/LayoutSyncDownTarget.java [hunk 18]
```diff
@@ -39,12 +39,13 @@
...
 /**
  * Sync down target for object layouts. This uses the '/ui-api/layout' API to fetch object layouts.
+ * The easiest way to use this sync target is through {@link com.salesforce.androidsdk.smartsync.manager.LayoutSyncManager}.
  *
  * @author bhariharan
  */
 public class LayoutSyncDownTarget extends SyncDownTarget {
 
-    public static final String SOBJECT_TYPE = "sobjectType";
+    public static final String OBJECT_TYPE = "objectType";
```
Comment (wmathurin) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/target/LayoutSyncDownTarget.java]: We are using sobjectType in MRUSyncDownTarget - should we stay consistent with that?

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+/**
+ * Provides an easy way to fetch layout data using {@link com.salesforce.androidsdk.smartsync.target.LayoutSyncDownTarget}.
+ * This class handles creating a soup, storing synched data and reading it into
+ * a meaningful data structure, i.e. {@link com.salesforce.androidsdk.smartsync.model.Layout}.
+ *
+ * @author bhariharan
+ */
+public class LayoutSyncManager {
+
+    private static final String SOUP_NAME = "layouts";
```
[the remark this answers is not in the document]
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Sure - I can change that.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java (hunk 14 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Okay - I can take a look at that.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java (hunk 16 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Nope - this is not our cache mode. This is the mode in the payload of the layout response itself, could be `Create`, `Update`, etc. It represents record display mode I think.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java (hunk 17 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Nope. This is coming from the response itself. Besides, we don't have an enum of layout types - we leave it up to the UI SDK doc for that with a link. That way more types can be added in the future and no client-side changes or mutations would be required.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/target/LayoutSyncDownTarget.java (hunk 18 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/target/LayoutSyncDownTarget.java]: Sure - I can change that. I changed it to match the `describe` response. I think that's `objectType`.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java (hunk 15 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Nope - this would probably happen when the server returned an error of some kind, or the object doesn't have a layout or something like that (I don't know if that's a real possibility).

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java (hunk 17 above)
Comment (wmathurin) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Is there a fixed set of possible values?

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java (hunk 16 above)
Comment (wmathurin) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: If there is a fixed set of possible values, it would be better represented with an enum.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java (hunk 15 above)
Comment (wmathurin) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Not sure I follow. If there is an error during sync down, nothing should be saved to the table so the first if should be entered.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java (hunk 17 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Right now it's `Compact` or `Full` I think, but it's not explicitly called out in the UI API doc, nor is it an `enum` on the server. Besides, the user already knows this since the user passes this into his/her call (`layoutType` is an argument). Right now, this class is a 1:1 representation of the layout response and uses the same data types as mentioned in the UI API doc. I'd prefer not to deviate from that. If the server side doc for the API was an `enum` of a specified set of values, we could change it, but for now, I feel like it's best to mirror the UI API doc 1:1.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java (hunk 16 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: The UI API doc does not call out whether it's a fixed set of values or not. I'd prefer to mirror the API response data model per UI API doc and avoid mutation on the client. If the server doc is updated to specify a set of possible values (`enum`), we can update this to `enum` at that time.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java (hunk 15 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: True. I guess this condition should never occur or would occur only if the response was a valid empty payload. Is that even possible?

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+ * Provides an easy way to fetch layout data using {@link com.salesforce.androidsdk.smartsync.target.LayoutSyncDownTarget}.
+ * This class handles creating a soup, storing synched data and reading it into
+ * a meaningful data structure, i.e. {@link com.salesforce.androidsdk.smartsync.model.Layout}.
+ *
+ * @author bhariharan
+ */
+public class LayoutSyncManager {
+
+    private static final String SOUP_NAME = "sfdcLayouts";
+    private static final String FEATURE_LAYOUT_SYNC = "LY";
```
Comment (bhariharan) [inline libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Also adding new feature flag for this.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java (hunk 17 above)
Comment (wmathurin) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/model/Layout.java]: Ok.

File: libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java (hunk 15 above)
Comment (bhariharan) [reply libs/SmartSync/src/com/salesforce/androidsdk/smartsync/manager/LayoutSyncManager.java]: Okay - will remove it.

**Comment by bhariharan** [comment]
Removed [here](https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/1695/commits/d9a39eea8d334f0395c5084fa74afb5472d3418e).

**Comment by wmathurin** [comment]
Another possible enhancement. 

Right now the config file that defines the syncs has only one top level key supported (it's "syncs"), for an example see: https://github.com/forcedotcom/SalesforceMobileSDK-iOS/blob/master/native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/usersyncs.json

Maybe we could support configuring meta-data synching right there.

For instance, you could have 
```javascript
{  
   "objectInfos":[  
      "Contact",
      "Account"
   ], // => setup object info sync for Contact and Account
   "layouts":[  
      "Contact":[  
         "Full"
      ], // => setup layout sync for Full layout for Contact
      "Account":[  
         "Compact",
         "Full"
      ] // => setup layout sync for Full and Compact layouts for Account
   ],
   "syncs":[  
      {  
         "syncName":"syncDownContacts",
         "syncType":"syncDown",
         "soupName":"contacts",
         "target":{  
            "type":"soql",
            "query":"SELECT FirstName, LastName, Title, MobilePhone, Email, Department, HomePhone FROM Contact LIMIT 10000"
         },
         "options":{  
            "mergeMode":"OVERWRITE"
         }
      }
   ]
}
```

Thoughts?

**Comment by bhariharan** [comment]
Good idea. We could do that. Should we evaluate usage of this and implement that as a 2nd iteration? ;-)

**Comment by bhariharan** [comment]
My next PR will have an equivalent helper for metadata.

**Comment by wmathurin** [comment]
Usage might be low because of discoverability - having it in the config might help on that front.
